builder/remotecontext/lazycontext.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/pkg/errors"
+)
+
+// NewLazyContext creates a new LazyContext. LazyContext defines a hashed build
+// context based on a root directory. Individual files are hashed first time
+// they are asked.
+func NewLazyContext(root string) (builder.Context, error) {
+	return &lazyContext{
+		root: root,
+		sums: make(map[string]string),
+	}, nil
+}
+
+type lazyContext struct {
+	root string
+	sums map[string]string
+}
+
+func (c *lazyContext) Close() error {
+	return nil
+}
+
+func (c *lazyContext) Open(path string) (io.ReadCloser, error) {
+	cleanPath, fullPath, err := c.normalize(path)
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := os.Open(fullPath)
+	if err != nil {
+		return nil, errors.WithStack(convertPathError(err, cleanPath))
+	}
+	return r, nil
+}
+
+func (c *lazyContext) Stat(path string) (string, builder.FileInfo, error) {
+	// TODO: although stat returns builder.FileInfo it builder.Context actually requires Hashed
+	cleanPath, fullPath, err := c.normalize(path)
+	if err != nil {
+		return "", nil, err
+	}
+
+	st, err := os.Lstat(fullPath)
+	if err != nil {
+		return "", nil, errors.WithStack(convertPathError(err, cleanPath))
+	}
+
+	relPath, err := rel(c.root, fullPath)
+	if err != nil {
+		return "", nil, convertPathError(err, cleanPath)
+	}
+
+	sum, ok := c.sums[relPath]
+	if !ok {
+		sum, err = c.prepareHash(relPath, st)
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	fi := &builder.HashedFileInfo{
+		builder.PathFileInfo{st, fullPath, filepath.Base(cleanPath)},
+		sum,
+	}
+	return relPath, fi, nil
+}
+
+func (c *lazyContext) Walk(root string, walkFn builder.WalkFunc) error {
+	_, fullPath, err := c.normalize(root)
+	if err != nil {
+		return err
+	}
+	return filepath.Walk(fullPath, func(fullPath string, fi os.FileInfo, err error) error {
+		relPath, err := rel(c.root, fullPath)
+		if err != nil {
+			return err
+		}
+		if relPath == "." {
+			return nil
+		}
+
+		sum, ok := c.sums[relPath]
+		if !ok {
+			sum, err = c.prepareHash(relPath, fi)
+			if err != nil {
+				return err
+			}
+		}
+
+		hfi := &builder.HashedFileInfo{
+			builder.PathFileInfo{FileInfo: fi, FilePath: fullPath},
+			sum,
+		}
+		if err := walkFn(relPath, hfi, nil); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+func (c *lazyContext) prepareHash(relPath string, fi os.FileInfo) (string, error) {
+	p := filepath.Join(c.root, relPath)
+	h, err := NewFileHash(p, relPath, fi)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create hash for %s", relPath)
+	}
+	if fi.Mode()&(os.ModeDir|os.ModeSymlink|os.ModeDevice|os.ModeSocket|os.ModeNamedPipe) == 0 && fi.Size() > 0 {
+		f, err := os.Open(p)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to open %s", relPath)
+		}
+		defer f.Close()
+		if _, err := pools.Copy(h, f); err != nil {
+			return "", errors.Wrapf(err, "failed to copy file data for %s", relPath)
+		}
+	}
+	sum := hex.EncodeToString(h.Sum(nil))
+	c.sums[relPath] = sum
```

Comment:
Should'nt we make it explicit in NewLazyContext comments then ? (something like "this context is not should not be accessed concurrently")